crates/extensions/c8y_http_proxy/src/actor.rs

```diff
@@ -162,80 +166,145 @@ impl C8YHttpProxyActor {
+        // build the request
+        let request_builder = build_request(&self.end_point);
+        let request = request_builder?.build()?;
+        // retry the request
         Ok(self.peers.http.await_response(request).await?)
     }
 
-    async fn create_event(
+    async fn try_request_with_fresh_internal_id(
         &mut self,
-        mut c8y_event: C8yCreateEvent,
-    ) -> Result<EventId, C8YRestError> {
-        if c8y_event.source.is_none() {
-            c8y_event.source = Some(C8yManagedObject {
-                id: self.end_point.get_c8y_internal_id().to_string(),
-            });
-        }
-        self.send_event_internal(c8y_event).await
+        device_type: DeviceType,
+        build_request: impl Fn(&C8yEndPoint) -> Result<HttpRequestBuilder, C8YRestError>,
+    ) -> Result<HttpResult, C8YRestError> {
+        // get new internal id not the cached one
+        self.get_and_set_internal_id(device_type).await?;
+
+        let request_builder = build_request(&self.end_point);
+        let request = request_builder?
+            .bearer_auth(self.end_point.token.clone().unwrap_or_default())
+            .build()?;
+        Ok(self.peers.http.await_response(request).await?)
+    }
+
+    async fn create_event(&mut self, c8y_event: C8yCreateEvent) -> Result<EventId, C8YRestError> {
+        let create_event = |internal_id: String| -> C8yCreateEvent {
+            C8yCreateEvent {
+                source: Some(C8yManagedObject { id: internal_id }),
+                event_type: c8y_event.event_type.clone(),
+                time: c8y_event.time,
+                text: c8y_event.text.clone(),
+                extras: c8y_event.extras.clone(),
+            }
+        };
+        let device_type = self.end_point.get_device_type(None);
+        self.send_event_internal(device_type, create_event).await
     }
 
     async fn send_software_list_http(
         &mut self,
         software_list: C8yUpdateSoftwareListResponse,
     ) -> Result<Unit, C8YRestError> {
-        let url = self.end_point.get_url_for_sw_list();
-        let req_builder = HttpRequestBuilder::put(url)
-            .header("Accept", "application/json")
-            .header("Content-Type", "application/json")
-            .json(&software_list);
-        let http_result = self.execute(req_builder).await?;
+        let build_request = |end_point: &C8yEndPoint| -> Result<HttpRequestBuilder, C8YRestError> {
+            let internal_id = end_point
+                .get_internal_id(DeviceType::MainDevice)
+                .map_err(|e| C8YRestError::CustomError(e.to_string()))?;
+            let url = end_point.get_url_for_sw_list(internal_id);
+            Ok(HttpRequestBuilder::put(url)
+                .header("Accept", "application/json")
+                .header("Content-Type", "application/json")
+                .json(&software_list))
+        };
+
+        let http_result = self.execute(DeviceType::MainDevice, build_request).await?;
```

Comment:
Here also, hardcoded to `MainDevice`. Something looks wrong here. I understand that we don't support software list upload for child devices ATM. But this generic API assuming the same seems wrong. This API should be able to send software list for any device.

Reply:
Thats right. Since the `Software list` is supported only by the main device, I have hardcoded here. When the software list feature is supported by the child device, then one has to pass the `device ID` to `send_software_list_http()`, and based on that URL can be created.